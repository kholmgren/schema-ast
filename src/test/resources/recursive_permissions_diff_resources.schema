definition tenant/organization {
	relation admin: tenant/user
	permission read_all_docs = admin
}

definition tenant/folder {
	relation parent: tenant/folder | tenant/organization
	relation reader: tenant/user

	// NOTE: since parent is either a tenant/folder OR an tenant/organization, if we want to check
	// in both, we need to make sure `read_all_docs` has the *same name* in both tenant/folder
	// and tenant/organization.
	permission read_all_docs = reader + parent->read_all_docs
}

definition tenant/document {
	relation parent: tenant/folder | tenant/organization
	relation reader: tenant/user

	permission view = reader + parent->read_all_docs
}

definition tenant/user {}
