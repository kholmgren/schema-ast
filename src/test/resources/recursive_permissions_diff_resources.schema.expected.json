{
  "@type" : "schema",
  "definitions" : [ {
    "@type" : "definition",
    "name" : "tenant/organization",
    "relations" : [ {
      "@type" : "relation",
      "name" : "admin",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read_all_docs",
      "expression" : {
        "@type" : "ref",
        "ref" : "admin"
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/folder",
    "relations" : [ {
      "@type" : "relation",
      "name" : "parent",
      "expression" : {
        "@type" : "or",
        "left" : {
          "@type" : "ref",
          "ref" : "tenant/folder"
        },
        "right" : {
          "@type" : "ref",
          "ref" : "tenant/organization"
        }
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read_all_docs",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "walk",
          "refs" : [ "parent", "read_all_docs" ]
        }
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/document",
    "relations" : [ {
      "@type" : "relation",
      "name" : "parent",
      "expression" : {
        "@type" : "or",
        "left" : {
          "@type" : "ref",
          "ref" : "tenant/folder"
        },
        "right" : {
          "@type" : "ref",
          "ref" : "tenant/organization"
        }
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "view",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "walk",
          "refs" : [ "parent", "read_all_docs" ]
        }
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/user"
  } ]
}